src/atama/head/head.cpp

```diff
@@ -0,0 +1,582 @@
+
+  switch (scan_direction) {
+    case 0:
+      scan_pan_angle += scan_speed;
+      if (scan_pan_angle > scan_left_limit) {
+        scan_position = (scan_position + 1) % 3;
+        scan_direction = 1;
+      }
+      break;
+    case 1:
+      scan_pan_angle -= scan_speed;
+      if (scan_pan_angle < scan_right_limit) {
+        scan_position = (scan_position + 1) % 3;
+        scan_direction = 0;
+      }
+      break;
+  }
+}
+
+void Head::scan_process()
+{
+  for (size_t i = 0; i < joints.size(); ++i) {
+    joints[i].set_pid_gain(20.0, 3.0, 2.0);
+  }
+
+  if (is_started_scanning) {
+    switch (scan_mode) {
+      case SCAN_UP:
+      case SCAN_DOWN:
+      case SCAN_CUSTOM: scan_two_direction(); break;
+      case SCAN_VERTICAL:
+      case SCAN_HORIZONTAL: scan_one_direction(); break;
+    }
+
+    pan_angle = pan_center + keisan::clamp(scan_pan_angle, right_limit, left_limit);
+    tilt_angle = tilt_center + keisan::clamp(scan_tilt_angle, bottom_limit, top_limit);
+  }
+
+  // set calculation result
+  if (!joints.empty() && joints.size() >= 2) {
+    joints[0].set_position(pan_angle);
+    joints[1].set_position(tilt_angle);
+  }
+}
+
+void Head::scan_custom(int scan_type)
+{
+  if (!check_time_belief()) {
+    return;
+  }
+
+  function_id = scan_type;
+  scan(scan_type);
+  scan_process();
+}
+
+double Head::calculate_distance_from_pan_tilt(double pan, double tilt)
+{
+  double distance = 0.0;
+
+  for (int i = 6; i >= 0; --i) {
+    for (int j = 0; j <= i; ++j) {
+      double x1 = pow((pan + pan_center + pan_offset), (i - j));
+      double x2 = pow((tilt + tilt_center + tilt_offset), j);
+      distance += (pan_tilt_to_distance_[i - j][j]) * x1 * x2;
+    }
+  }
+
+  return (distance > 0.0) ? distance : 0.0;
+}
+
+double Head::calculate_distance_from_tilt(double tilt)
+{
+  double distance = 0.0;
+
+  for (int i = 0; i <= 4; ++i) {
+    double x2 = pow((tilt + tilt_center + tilt_offset), i);
+    distance += (tilt_to_distance_[i]) * x2;
+  }
+
+  return (distance > 0.0) ? distance : 0.0;
+}
+
+double Head::calculate_tilt_from_pan_distance(double pan, double distance)
+{
+  double tilt = 0.0;
+
+  for (int i = 4; i >= 0; --i) {
+    for (int j = 0; j <= i; ++j) {
+      double x1 = pow((pan + pan_center + pan_offset), (i - j));
+      double x2 = pow(distance, j);
+      tilt += (pan_distance_to_tilt_[i - j][j]) * x1 * x2;
+    }
+  }
+
+  return ((tilt < 15.0) ? tilt : 15.0) - tilt_center - tilt_offset;
+}
+
+void Head::look_to_position(
+  double goal_position_x, double goal_position_y,
+  double robot_position_x, double robot_position_y,
+  float yaw)
+{
+  function_id = Head::LOOK_TO_POSITION;
+  float dx = goal_position_x - robot_position_x;
+  float dy = goal_position_y - robot_position_y;
+
+  float pan = yaw -
+    keisan::make_degree(keisan::signed_arctan(dy, dx).degree()).normalize().degree();
+  float tilt = calculate_tilt_from_pan_distance(
+    keisan::make_degree(keisan::signed_arctan(dy, dx).degree()).degree());
```

Comment:
`keisan::signed_arctan` already return `keisan::Angle`, so no need to create that again, just do this `keisan::signed_arctan().normalize().degree()`

Reply:
ok